core/types/block.go

```diff
@@ -398,23 +398,45 @@ func (b *Block) Hash() common.Hash {
 }
 
 // L1MessageCount returns the number of L1 messages in this block.
+// This count also includes skipped messages.
 func (b *Block) L1MessageCount() int {
 	if l1MsgCount := b.l1MsgCount.Load(); l1MsgCount != nil {
 		return l1MsgCount.(int)
 	}
-	count := 0
-	for _, tx := range b.transactions {
-		if tx.IsL1MessageTx() {
-			count += 1
+
+	// find first and last queue index in block
+	var firstQueueIndex *uint64
+	var lastQueueIndex *uint64
+
+	for ii, tx := range b.transactions {
+		if !tx.IsL1MessageTx() {
+			break
+		}
+		if ii == 0 {
+			firstQueueIndex = &b.transactions[ii].AsL1MessageTx().QueueIndex
 		}
+		lastQueueIndex = &b.transactions[ii].AsL1MessageTx().QueueIndex
+	}
+
+	// calculate and cache L1 message count
+	count := 0
+	if firstQueueIndex != nil {
+		// lastQueueIndex is guaranteed to be non-nil in this case
+		count = int(*lastQueueIndex - *firstQueueIndex + 1)
```

Comment:
I think it is not correct. Consider the situation below:

+ block 0: contains message 0 to 5
+ block 1: contains message 10 to 15

The l1 message count (including skipped messages) for block 0 is 6, but for block 1 the l1 message count (including skipped messages) is 10.

Reply:
True, seems like we need to pass additional context here.